### PR TITLE
[flow] Switch to directional shadows

### PR DESCRIPTION
**Pull request by untp**
Shadows incorrectly offset, shifts incrementally. This issue fixed by Skia with [Issue 10781: Add orthographic shadow support](https://bugs.chromium.org/p/skia/issues/detail?id=10781). We need to add `SkShadowFlags::kDirectionalLight_ShadowFlag` to `DrawShadow`.

## Related Issues
fixes flutter/flutter#28463
fixes flutter/flutter#51237

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: flow/layers/physical_shape_layer.cc
Comment (flar) [inline flow/layers/physical_shape_layer.cc:20]: The code used to use an offset of -600 from the top of the bounds, but the value is now hard-coded as a -450 offset directionally. Doesn't that produce a different effect?

Comment (flar) [reply flow/layers/physical_shape_layer.cc:20]: Or is this change intentional due to undesired results of the previous fixed offset?

File: flow/layers/physical_shape_layer.cc
Comment (flar) [inline flow/layers/physical_shape_layer.cc:147]: I'm glad to see this switch to using the Shadow utils. I discovered that the previous implementation of PSL:CSB was not producing the right values when I was doing bounds testing for the new DisplayList code, but hadn't had a chance to come back and see what the problem was.

A descriptive comment here on why and where the DPR is being factored in would be nice. In my brief investigation I could see that the matrix had DPR built in for X and Y, but there was no scale for Z so the code here seems to be manually scaling the Z components of the lighting equations by that factor. One question is that the LightOffset parameters are not being scaled by anything, but perhaps the ShadowUtils will interpret them relative to the CTM being supplied. Is that then consistent with modifying kLightHeight by DPR, but not the other values? Also, the light radius is in the X & Y directions, so does it produce the right behavior when multiplied by DPR?

Have you run tests that show that these equations produce the right values as the widget is scaled? (in other words, the shadow grows or does not grow as expected?)

Comment (flar) [reply flow/layers/physical_shape_layer.cc:147]: Also, if you are rebased to a recent ToT then you should see `flow/display_list_utils.cc` with calculations to compute the bounds of a shadow. It would be nice to double check it and make it consistent with the work being done here.

File: flow/layers/physical_shape_layer_unittests.cc
Comment (flar) [inline flow/layers/physical_shape_layer_unittests.cc:178]: I'd like to see some new tests written that verify that we produce the expected bounds for a given set of parameters.

**Review by flar — CHANGES_REQUESTED** [review]

File: flow/layers/physical_shape_layer.cc
Comment (untp) [reply flow/layers/physical_shape_layer.cc:20]: I don't know. I just copied these values from [canvaskit/util.dart](https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/canvaskit/util.dart). We should ask this to @yjbanov.

File: flow/layers/physical_shape_layer.cc
Comment (untp) [reply flow/layers/physical_shape_layer.cc:147]: I didn't thought this. I just copied the old implementation. Now I'm going to test that `ComputeShadowBounds` calculates the right values between different `dpr` values, and add that test to unit tests.

Also I will update `display_list_utils.cc`.

File: flow/layers/physical_shape_layer_unittests.cc
Comment (untp) [reply flow/layers/physical_shape_layer_unittests.cc:178]: Yes, testing with different values of `dpr` and `ctm` will be nice. Working on it.

**Comment by untp** [comment]
I investigated `GetLocalBounds` for compatibility with old implementation.
Getting the same results of the old implementation is impossible.
So we need to choose values that gives a result closest to the old implementation.
Here are the simplified equations from `GetLocalBounds`:
```python
ambientBlur = 0.5 * zPlane.z / ctm.getMinScale()
if kDirectionalLight_ShadowFlag in flags:
  spotBlur = lightRadius * zPlane.z / ctm.getMinScale()
  spotScale = 1
  spotOffset.x = - lightPos.normalize().x * zPlane.z / ctm.getMinScale()
  spotOffset.y = - lightPos.normalize().y * zPlane.z / ctm.getMinScale()
else:
  spotBlur = lightRadius * (zPlane.z / (lightPos.z - zPlane.z)) / ctm.getMinScale()
  spotScale = lightPos.z / (lightPos.z - zPlane.z)
  spotOffset.x = - (zPlane.z / (lightPos.z - zPlane.z)) *
                       ((lightPos.x - ctm.getTranslateX()) / ctm.getMinScale())
  spotOffset.y = - (zPlane.z / (lightPos.z - zPlane.z)) *
                       ((lightPos.y - ctm.getTranslateY()) / ctm.getMinScale())
```
If these equations applied to the old implementation:
(assume `ctm.getMinScale()` equal to `dpr`)
```python
ambientBlur = 0.5 * elevation
spotBlur = kLightRadius * elevation / (kLightHeight - elevation)
spotScale = kLightHeight / (kLightHeight - elevation)
spotOffset.x = - elevation / (kLightHeight - elevation) *
    (((bounds.left() + bounds.right()) / 2 - ctm.getTranslateX()) / ctm.getMinScale())
spotOffset.y = - elevation / (kLightHeight - elevation) *
    ((bounds.top() - kLightHeight - ctm.getTranslateY()) / ctm.getMinScale())
```
These 5 variables needs to be equal between old and new implementation but it is impossible.
`spotScale` can not be equal because `1 != kLightHeight / (kLightHeight - elevation)`.
If we assume `kLightHeight - elevation` equals to `kLightHeight`, variables become:
(Also assuming `ctm.getTranslateX() == (bounds.left() + bounds.right()) / 2`
and assuming `ctm.getTranslateY() == bounds.top()`)
```python
ambientBlur = 0.5 * elevation
spotBlur = kLightRadius / kLightHeight * elevation
spotScale = 1
spotOffset.x = 0
spotOffset.y = elevation / ctm.getMinScale()
```
Let's find out new implementation's values:
```python
0.5 * zPlane.z / ctm.getMinScale() = 0.5 * elevation
lightRadius * zPlane.z / ctm.getMinScale() = kLightRadius / kLightHeight * elevation
- zPlane.z * lightPos.normalize().x / ctm.getMinScale() = 0
- zPlane.z * lightPos.normalize().y / ctm.getMinScale() = elevation / ctm.getMinScale()
```
After simplifying:
```python
zPlane.z = elevation * ctm.getMinScale()
lightRadius = kLightRadius / kLightHeight
lightPos.normalize().x = 0
lightPos.normalize().y = - 1 / ctm.getMinScale()
```
`ctm.getMinScale()` mostly greater than 1 but can be smaller than 1 on low resolution devices.
And normalize() can't return smaller than -1; so, we need to assume `ctm.getMinScale()` is 1 only for `lightPos.normalize().y`.
Therefore we set `lightPos` as `(0, -1, 0)`.
So function call should be like this:
(Using `dpr` as `ctm.getMinScale()`, because sometimes `ctm.getMinScale() != dpr`)
```c++
SkShadowUtils::GetLocalBounds(
    ctm,
    path,
    SkPoint3::Make(0, 0, elevation * dpr),
    SkPoint3::Make(0, -1, 0),
    kLightRadius / kLightHeight,
    SkShadowFlags::kDirectionalLight_ShadowFlag,
    &shadow_bounds);
```
These are results of different rects calculated with old and new implementation:
(elevation = 1, dpr = 1, ctm = SkMatrix())
```
--------------------------------
rect: LTRB: -1, 0, 1, 1
old_impl: LTRB: -3.33723, -1.5, 3.33723, 4.3389
new_impl: LTRB: -3.33333, -1.5, 3.33333, 4.33333
--------------------------------
rect: LTRB: -10, 0, 10, 10
old_impl: LTRB: -12.3523, -1.5, 12.3523, 13.3539
new_impl: LTRB: -12.3333, -1.5, 12.3333, 13.3333
--------------------------------
rect: LTRB: -100, 0, 100, 100
old_impl: LTRB: -102.502, -1.5, 102.502, 103.504
new_impl: LTRB: -102.333, -1.5, 102.333, 103.333
--------------------------------
rect: LTRB: -1, 0, 1, 2
old_impl: LTRB: -3.33723, -1.5, 3.33723, 5.34057
new_impl: LTRB: -3.33333, -1.5, 3.33333, 5.33333
--------------------------------
rect: LTRB: -10, 0, 10, 20
old_impl: LTRB: -12.3523, -1.5, 12.3523, 23.3706
new_impl: LTRB: -12.3333, -1.5, 12.3333, 23.3333
--------------------------------
rect: LTRB: -100, 0, 100, 200
old_impl: LTRB: -102.502, -1.5, 102.502, 203.671
new_impl: LTRB: -102.333, -1.5, 102.333, 203.333
--------------------------------
rect: LTRB: 0, 0, 1, 1
old_impl: LTRB: -2.33639, -1.5, 3.33639, 4.3389
new_impl: LTRB: -2.33333, -1.5, 3.33333, 4.33333
--------------------------------
rect: LTRB: 0, 0, 10, 10
old_impl: LTRB: -2.34391, -1.5, 12.3439, 13.3539
new_impl: LTRB: -2.33333, -1.5, 12.3333, 13.3333
--------------------------------
rect: LTRB: 0, 0, 100, 100
old_impl: LTRB: -2.41903, -1.5, 102.419, 103.504
new_impl: LTRB: -2.33333, -1.5, 102.333, 103.333
--------------------------------
rect: LTRB: 1, 1, 2, 2
old_impl: LTRB: -1.33639, -0.5, 4.33639, 5.3389
new_impl: LTRB: -1.33333, -0.5, 4.33333, 5.33333
--------------------------------
rect: LTRB: 10, 10, 20, 20
old_impl: LTRB: 7.65609, 8.5, 22.3439, 23.3539
new_impl: LTRB: 7.66667, 8.5, 22.3333, 23.3333
--------------------------------
rect: LTRB: 100, 100, 200, 200
old_impl: LTRB: 97.581, 98.5, 202.419, 203.504
new_impl: LTRB: 97.6667, 98.5, 202.333, 203.333
--------------------------------
```

**Comment by flar** [comment]
> I investigated `GetLocalBounds` for compatibility with old implementation.

I think that may be a mistake in tactics. This assumes that the old implementation was correct. In particular, when I was using the PhysicalShapeLayer to compute bounds for my DisplayList tests I ran into problems with it not enclosing all of the pixels that were rendered (the DL tests do OOB testing for rendering vs DL.bounds()). I ended up ignoring it and using GetLocalBounds in the shadow utils instead for DL bounds computations.

The goal here will be to make sure that the bounds computed encompass the rendering, not that they are compatible with the prior implementation of something. It would be nice if they were no looser than the old implementation as well, for performance.

With respect to rendering, we need to decide what the goal is here. Some of the comments linked above talked about not matching the Material spec, and I think we do need to ensure our Material widgets match the material spec.

We also need to consider that we have Cupertino widgets that may also use these shadows and need to be compatible with the Cupertino spec.

If they are dragging us in 2 different directions with how we interpret various values like elevation and light direction, then maybe we need to add some extra parameters to our shadow parameters, but hopefully they have similar enough shadow specs that we can meet them with the existing parameters.

**Comment by untp** [comment]
I think, people want to see same shadows after this fix. Because this PR edits code that 4 years old. Material specification is not strict, it doesn't provide values, it just say [combine shadow from key and ambient lights](https://material.io/design/environment/light-shadows.html).

I don't know what is the goal. I just fixed the above issues. Now shadows doesn't depend on `ctm`.

Maybe we need to invite someone from material design team.

Also here are some screenshots
<details><summary>Old implementation</summary>

old

</details>

<details><summary>New implementation</summary>

new

</details>

**Comment by flar** [comment]
I tagged @HansMuller who can weigh in on what Material's requirements are for shadows and coordinate transforms, or point to someone who would know.

File: flow/layers/physical_shape_layer.cc
Comment (yjbanov) [reply flow/layers/physical_shape_layer.cc]: The values we use on the web side don't have much science to them. We just eyeballed the result and checked with the material team. In the 6 months since the change we've received zero complaints about how the shadows look.

**Comment by yjbanov** [comment]
If it helps, here's how we [test shadows](https://github.com/flutter/engine/blob/5c9e305196dbc28477cff0f1c915e0bf92bf316c/lib/web_ui/test/canvaskit/canvas_golden_test.dart#L94) ([golden](https://github.com/flutter/goldens/blob/master/engine/web/canvaskit_directional_shadows.png)) and [shadow bounds](https://github.com/flutter/engine/blob/5c9e305196dbc28477cff0f1c915e0bf92bf316c/lib/web_ui/test/canvaskit/canvas_golden_test.dart#L154) ([golden](https://github.com/flutter/goldens/blob/master/engine/web/canvaskit_shadow_bounds.png)) on the web.

I recommend that we prioritize fixing the egregiousness that's going on in our current shadow implementation, and fine-tune the shadows for spec compliance later. Shadows are unusable on desktop (and to a large extent on tablets). I think Cupertino isn't affected because HIG doesn't use shadows (although perhaps some of our Cupertino widgets do?).

**Review by yjbanov — APPROVED** [review]
LGTM if LGT @flar and @chinmaygarde

File: flow/layers/physical_shape_layer_unittests.cc
Comment (flar) [inline flow/layers/physical_shape_layer_unittests.cc:204]: The original bugs seemed to be concerned about the look of the shadow as the size of the widget changed. It would be nice to test that.

Perhaps either some golden tests for shadows of various sizes of path, or a rendering test that draws shadows of 2 paths of different sizes, but with the top extending out of view so that the part visible in the generated image should look identical?

**Review by flar — CHANGES_REQUESTED** [review]

**Comment by flar** [comment]
So, I can see how the previous implementation created lots of problems by having the shadow position relative to the bounds, and I agree that a directional shadow would likely match developers expectations better.

Have any or all of the test cases in the 2 attached issues been run with the fix to see if they behave according to expectations?

File: flow/layers/physical_shape_layer_unittests.cc
Comment (untp) [reply flow/layers/physical_shape_layer_unittests.cc:204]: That bug is not reproducible on latest version (https://github.com/flutter/flutter/issues/28463#issuecomment-762756193). But I can create a test with different sizes using `PhysicalShapeLayer::ComputeShadowBounds`. We can substract shadow bounds and path bounds, then checking them equal across different sizes (with same elevation). I don't want to add a golden test (Because I don't know).

**Comment by untp** [comment]
For issue flutter/flutter#28463, I created this comment (https://github.com/flutter/flutter/issues/28463#issuecomment-872578526), shadows looked perfect.
I also run this [code](https://gist.github.com/rydmike/145828269bd8d24ee9c44a9df26ec7fb) from https://github.com/flutter/flutter/issues/51237#issuecomment-670263611. Everything looked good.

File: flow/layers/physical_shape_layer_unittests.cc
Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc:204]: A test that rendered 2 shadows with paths of different sizes and expected the same size shadow at the bottom should be correct even if the values might eventually be tweaked for compliance with a spec. The point is how the shadow reacts to the path size and the answer should be "it shouldn't" if I am reading all of this right?

Something like:

Record a picture of a shadow of a rect LTRB(20, -50, 80, 80). Rasterize it to an image that is (0,0,100, >100) (large enough to capture the shadow below the rect, but not the top of the rect). Only the bottom of the rect and shadow should show.
Record the same thing, but the start Y of the rect is -500 or something larger, it still ends at Y=80 per the previous image.
The two images should be identical.

Do the same thing for the top of a small and large rect (have it extend off the bottom of the image this time).
Those two images should be identical.

The display_list_canvas_unittests.cc have code that renders a bunch of primitives and checks their output against each other that you can lift.

Even if we eventually tweak the values to match a spec, these tests still verify that the shadows above and below a path are unchanged by the size of the path.

Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc:204]: I reference this method in the display_list_canvas_unittests.cc. It's not testing whether the rendering is correct, but just that we render the same thing via direct calls to a surface canvas and indirectly through a DisplayList. If Skia fixes a bug that makes the rendering different, the same difference should appear in both images and still compare ==. Similarly, if we ever change how big or dark the shadow is to match a Material or Cupertino spec, we can still have a test that makes sure that the size of the "card list" or path doesn't affect the size or look of the shadow.

https://github.com/flutter/engine/blob/b57c5025159ba6f3c5703e2a6d2e433e4809bfcc/flow/display_list_canvas_unittests.cc#L666

Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc:204]: You can ignore the additional bounds testing in those cases, I was testing both the accuracy of the bounds from both mechanisms as well as the identical output. I was also testing that "something is rendered" (i.e. pixels_touched > 0) because I wanted to make sure that none of my tests requested a NOP rendering, but that isn't a critical part of the test.

**Comment by flar** [comment]
> For issue [flutter/flutter#28463](https://github.com/flutter/flutter/issues/28463), I created this comment ([flutter/flutter#28463 (comment)](https://github.com/flutter/flutter/issues/28463#issuecomment-872578526)), shadows looked perfect.
> I also run this [code](https://gist.github.com/rydmike/145828269bd8d24ee9c44a9df26ec7fb) from [flutter/flutter#51237 (comment)](https://github.com/flutter/flutter/issues/51237#issuecomment-670263611). Everything looked good.

That's a great idea. It keeps the original issue filers apprised of how things will change for them.

File: flow/layers/physical_shape_layer_unittests.cc
Comment (untp) [reply flow/layers/physical_shape_layer_unittests.cc:204]: Thanks for detailed tutorial @flar. I created the test.

File: flow/layers/physical_shape_layer_unittests.cc
Comment (flar) [inline flow/layers/physical_shape_layer_unittests.cc:261]: Perhaps use `return pixels_different` (with a different function name) so that the caller can `EXPECT_EQ(..., 0)` and we'll see just how many pixels failed in the error logs.

File: flow/layers/physical_shape_layer_unittests.cc
Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc:204]: NP, we should probably add a test utility to do things like this, but don't worry about that for now. I'll take that as an action item.

**Review by flar — COMMENTED** [review]
The outstanding code comments are basically a naming/return value nit about the new test which you can take or leave as you see fit.

I'm going to ask someone from Skia to look this over as well. I read through the math and think it all makes sense, but it seems like we are driving this from the math rather than from what we want to accomplish and I want to see if we've missed something more basic.

**Comment by flar** [comment]
@jvanverth said he would look over the math and API usage here to see if the new code is as they intended for the new directional shadows.

**Comment by flar** [comment]
It might be a day or two until we can get some attention from the busy Skia team. On the surface, it sounds like they are concerned as I was that our light direction had a Z component of 0 since that would imply that the light was coplanar (with...?), so I think their review might be informative.

**Comment by flar** [comment]
Just a note that when this goes live we are going to have a ton of golden file failures throughout the system to verify and approve. That mass update of goldens is one reason that I want to confirm with Skia that we are "using it right". I'd hate to do all that work and then find out that there was something going not by design here and have to potentially introduce another mass change of every golden test that uses shadows.

**Comment by jvanverth** [comment]
So the reason the resulting direction is unintuitive is that this case has exposed a bug in the directional lighting code. Rather than 
```
spotOffset.x = - lightPos.normalize().x * zPlane.z / ctm.getMinScale()
spotOffset.y = - lightPos.normalize().y * zPlane.z / ctm.getMinScale()
```
it should be 
```
spotOffset.x = - lightPos.x / lightPos.z * zPlane.z / ctm.getMinScale()
spotOffset.y = - lightPos.y / lightPos.z * zPlane.z / ctm.getMinScale()
```
(This also makes me think that the blur could be
```
spotBlur = lightRadius / lightPos.z * zPlane.z / ctm.getMinScale()
```
though either way it's sort of a hack. A true directional light wouldn't normally affect blur, since it's infinitely far away.)

In any case, with those changes `lightPos` should end up being `(0, sqrt(2)/2, sqrt(2)/2`, which makes more sense as a light direction.

**Comment by jvanverth** [comment]
Oh, and once I land this fix it probably means that the Web values would have to change, depending on how sharp the lighting angle is.

**Comment by flar** [comment]
> In any case, with those changes `lightPos` should end up being `(0, -sqrt(2)/2, sqrt(2)/2`, which makes more sense as a light direction.

Wouldn't `(0, -1, 1)` work as well since the values are "Z-normalized"?

**Comment by jvanverth** [comment]
Yeah, any vector pointing in that direction would work, I just normalized it out of habit.

**Comment by yjbanov** [comment]
> Oh, and once I land this fix it probably means that the Web values would have to change, depending on how sharp the lighting angle is.

We can address the web side in a separate PR. We roll Skia with CanvasKit releases, so this won't take effect immediately. Thanks for the heads up though.

**Comment by jvanverth** [comment]
> We can address the web side in a separate PR. We roll Skia with CanvasKit releases, so this won't take effect immediately. Thanks for the heads up though.

Sounds good, I'll land the fix tomorrow morning.

**Comment by flar** [comment]
When the Skia fix lands and this PR will have to be rebased to catch it, I think that may also fix the odd error in the Linux Android Scenarios as well.

**Comment by flar** [comment]
Just a status update. I'm on the CC thread for the Skia fix to the directional shadow math and it is actively being worked on. The delays are minor and related to making sure it passes all tests and golden image comparisons.

**Comment by jvanverth** [comment]
The fix has landed: https://skia-review.googlesource.com/c/skia/+/428978

**Comment by flar** [comment]
Try rebasing to >= https://github.com/flutter/engine/commit/c120a50ecb3b32927a387900418a6d09dfe1e287 to check with the fix in place.

**Comment by untp** [comment]
I already rebased with previous commits. My branch's [DEPS](https://github.com/untp/engine/blob/untp-patch-1/DEPS#L30)' skia_revision is `94fda947`.

**Review by flar — APPROVED** [review]
LGTM

**Pull request merged**
